File: extensions/cade-vscode/src/connection.ts
/**
 * CadeConnection — manages the TCP connection to `cade-ide-mcp`.
 *
 * ## Lifecycle
 *
 * 1. `connect()` — reads `~/.cade/ide/<pid-of-cade-ide-mcp>.json`,
 *    opens a TCP socket to the advertised address, exchanges
 *    `Hello` / `HelloAck`, then stays connected.
 * 2. Incoming `CallbackRequest` frames are dispatched to the handler
 *    registered via `onCallbackRequest()`.
 * 3. `sendStateUpdate()` encodes a `StateSnapshot` and writes it to
 *    the socket.
 * 4. `sendResponse()` encodes a `CallbackResponse` and writes it.
 * 5. `dispose()` destroys the socket and cancels reconnection.
 *
 * The connection performs a single reconnect attempt after a 3-second
 * delay when the socket closes unexpectedly.
 */

import * as fs from "fs";
import * as net from "net";
import * as os from "os";
import * as path from "path";
import * as readline from "readline";
import * as vscode from "vscode";

import {
  AdapterMessage,
  CallbackOp,
  CallbackResult,
  ServerMessage,
  StateSnapshot,
  decodeServerMessage,
  encodeAdapterMessage,
} from "./protocol";

const PROTOCOL_VERSION = 1;
const RECONNECT_DELAY_MS = 3_000;

/** Shape of the discovery JSON written by `cade-ide-mcp`. */
interface DiscoveryInfo {
  pid: number;
  addr: string; // "127.0.0.1:<port>"
}

export type CallbackRequestHandler = (
  id: number,
  op: CallbackOp,
) => void;

export class CadeConnection {
  private socket: net.Socket | undefined;
  private rl: readline.Interface | undefined;
  private disposed = false;
  private reconnectTimer: ReturnType<typeof setTimeout> | undefined;
  private _onCallbackRequest: CallbackRequestHandler | undefined;

  constructor(private readonly output: vscode.OutputChannel) {}

  /** Register the handler called when a `CallbackRequest` arrives. */
  onCallbackRequest(handler: CallbackRequestHandler): void {
    this._onCallbackRequest = handler;
  }

  /** Find and connect to the running `cade-ide-mcp` process. */
  async connect(): Promise<void> {
    if (this.disposed) return;
    this.clearSocket();

    const info = this.readDiscoveryFile();
    if (!info) {
      this.output.appendLine(
        "CADE: cade-ide-mcp not running (no discovery file found). " +
          `Retrying in ${RECONNECT_DELAY_MS / 1000}s…`,
      );
      this.scheduleReconnect();
      return;
    }

    const [host, portStr] = info.addr.split(":");
    const port = parseInt(portStr, 10);

    return new Promise<void>((resolve) => {
      const sock = new net.Socket();
      this.socket = sock;

      sock.once("connect", () => {
        this.output.appendLine(`CADE: connected to cade-ide-mcp at ${info.addr}`);
        this.setupReader(sock);
        this.sendHello(sock);
        resolve();
      });

      sock.once("error", (err) => {
        this.output.appendLine(`CADE: connection error — ${err.message}`);
        this.scheduleReconnect();
        resolve();
      });

      sock.once("close", () => {
        if (!this.disposed) {
          this.output.appendLine("CADE: disconnected from cade-ide-mcp. Reconnecting…");
          this.scheduleReconnect();
        }
      });

      sock.connect(port, host);
    });
  }

  /** Send a full state snapshot to the server. */
  sendStateUpdate(snapshot: StateSnapshot): void {
    const msg: AdapterMessage = { type: "state_update", ...snapshot };
    this.write(encodeAdapterMessage(msg));
  }

  /** Send a `CallbackResponse` for the given request id. */
  sendResponse(id: number, result: CallbackResult): void {
    const msg: AdapterMessage = { type: "callback_response", id, result };
    this.write(encodeAdapterMessage(msg));
  }

  dispose(): void {
    this.disposed = true;
    this.clearReconnectTimer();
    this.clearSocket();
  }

  // ── private ──────────────────────────────────────────────────────────────

  private sendHello(sock: net.Socket): void {
    const msg: AdapterMessage = {
      type: "hello",
      label: `vscode-${vscode.env.appName}`,
      protocol_version: PROTOCOL_VERSION,
    };
    sock.write(encodeAdapterMessage(msg));
  }

  private setupReader(sock: net.Socket): void {
    this.rl?.close();
    const rl = readline.createInterface({ input: sock, crlfDelay: Infinity });
    this.rl = rl;

    rl.on("line", (line) => {
      if (!line.trim()) return;
      let msg: ServerMessage;
      try {
        msg = decodeServerMessage(line);
      } catch (e) {
        this.output.appendLine(`CADE: malformed frame — ${String(e)}`);
        return;
      }
      this.handleServerMessage(msg);
    });
  }

  private handleServerMessage(msg: ServerMessage): void {
    if (msg.type === "hello_ack") {
      this.output.appendLine(
        `CADE: HelloAck received (protocol v${msg.protocol_version}). Adapter ready.`,
      );
      return;
    }
    if (msg.type === "callback_request") {
      const { id, type: _t, ...rest } = msg as ServerMessage & {
        type: "callback_request";
        id: number;
      };
      const op = rest as unknown as CallbackOp;
      this._onCallbackRequest?.(id, op);
    }
  }

  private write(data: string): void {
    if (this.socket && !this.socket.destroyed) {
      this.socket.write(data);
    }
  }

  private clearSocket(): void {
    this.rl?.close();
    this.rl = undefined;
    this.socket?.destroy();
    this.socket = undefined;
  }

  private scheduleReconnect(): void {
    if (this.disposed) return;
    this.clearReconnectTimer();
    this.reconnectTimer = setTimeout(() => {
      void this.connect();
    }, RECONNECT_DELAY_MS);
  }

  private clearReconnectTimer(): void {
    if (this.reconnectTimer !== undefined) {
      clearTimeout(this.reconnectTimer);
      this.reconnectTimer = undefined;
    }
  }

  private readDiscoveryFile(): DiscoveryInfo | undefined {
    const dir = path.join(os.homedir(), ".cade", "ide");
    let entries: fs.Dirent[];
    try {
      entries = fs.readdirSync(dir, { withFileTypes: true });
    } catch {
      return undefined;
    }

    // Pick the most-recently-modified discovery file — the running instance.
    const jsonFiles = entries
      .filter((e) => e.isFile() && e.name.endsWith(".json"))
      .map((e) => {
        const fullPath = path.join(dir, e.name);
        try {
          const stat = fs.statSync(fullPath);
          return { fullPath, mtimeMs: stat.mtimeMs };
        } catch {
          return null;
        }
      })
      .filter((x): x is { fullPath: string; mtimeMs: number } => x !== null)
      .sort((a, b) => b.mtimeMs - a.mtimeMs);

    if (jsonFiles.length === 0) return undefined;

    try {
      const raw = fs.readFileSync(jsonFiles[0].fullPath, "utf8");
      return JSON.parse(raw) as DiscoveryInfo;
    } catch {
      return undefined;
    }
  }
}

File: extensions/cade-neovim/lua/cade_ide/connection.lua
-- lua/cade_ide/connection.lua
-- Manages the TCP connection to cade-ide-mcp.
--
-- Architecture:
--   1. connect()   — reads ~/.cade/ide/<pid>.json, opens TCP via vim.loop,
--                    sends Hello, waits for HelloAck, then pumps the read loop.
--   2. Incoming frames are line-buffered; each complete line is decoded and
--      dispatched: HelloAck → log, CallbackRequest → on_callback handler.
--   3. send_state_update(snap)  — encodes + writes StateUpdate frame.
--   4. send_response(id, result)— encodes + writes CallbackResponse frame.
--   5. dispose()   — closes socket, cancels reconnect timer.
--
-- The public API is a table returned by M.new().

local proto = require("cade_ide.protocol")

local PROTOCOL_VERSION = 1
local RECONNECT_MS     = 3000
local LABEL            = "neovim-" .. (vim.version and
                           ("%d.%d"):format(vim.version().major, vim.version().minor)
                           or "unknown")
local DEFAULT_DISC_DIR = vim.fn.expand("~/.cade/ide")

local M = {}

--- Create a new connection object.
--- @param opts table|nil  { log = function(msg), discovery_dir = string }
function M.new(opts)
  opts = opts or {}
  local log = opts.log or function(msg)
    vim.schedule(function()
      vim.notify("[cade-ide] " .. msg, vim.log.levels.INFO)
    end)
  end
  local disc_dir = opts.discovery_dir or DEFAULT_DISC_DIR

  local self = {
    _tcp        = nil,
    _buf        = "",        -- incomplete line buffer
    _timer      = nil,
    _disposed   = false,
    _handler    = nil,       -- on_callback(id, op)
    log         = log,
  }

  -- ── public API ────────────────────────────────────────────────────────────

  function self.on_callback(handler) self._handler = handler end

  function self.connect()
    if self._disposed then return end
    self:_disconnect()

    local info = self:_read_discovery()
    if not info then
      self.log("cade-ide-mcp not running (no discovery file). Retrying…")
      self:_schedule_reconnect()
      return
    end

    local tcp = vim.loop.new_tcp()
    self._tcp = tcp
    self._buf = ""

    tcp:connect(info.addr, info.port, function(err)
      if err then
        self.log("connection failed: " .. err)
        self:_schedule_reconnect()
        return
      end

      -- Send Hello.
      self:_write(proto.encode(proto.hello(LABEL)))
      self.log(("connected to cade-ide-mcp at %s:%d"):format(info.addr, info.port))

      -- Start read loop.
      tcp:read_start(function(rerr, data)
        if rerr or not data then
          if not self._disposed then
            self.log("disconnected — reconnecting…")
            self:_schedule_reconnect()
          end
          return
        end
        self:_on_data(data)
      end)
    end)
  end

  function self.send_state_update(snap)
    self:_write(proto.encode(proto.state_update(snap)))
  end

  function self.send_response(id, result)
    self:_write(proto.encode(proto.callback_response(id, result)))
  end

  function self.dispose()
    self._disposed = true
    self:_cancel_timer()
    self:_disconnect()
  end

  -- ── private ───────────────────────────────────────────────────────────────

  function self:_write(data)
    if self._tcp and not self._tcp:is_closing() then
      self._tcp:write(data)
    end
  end

  function self:_disconnect()
    self:_cancel_timer()
    if self._tcp then
      if not self._tcp:is_closing() then
        pcall(function() self._tcp:read_stop() end)
        pcall(function() self._tcp:close() end)
      end
      self._tcp = nil
    end
    self._buf = ""
  end

  function self:_cancel_timer()
    if self._timer then
      pcall(function() self._timer:stop(); self._timer:close() end)
      self._timer = nil
    end
  end

  function self:_schedule_reconnect()
    if self._disposed then return end
    self:_cancel_timer()
    local t = vim.loop.new_timer()
    self._timer = t
    t:start(RECONNECT_MS, 0, function()
      t:close(); self._timer = nil
      if not self._disposed then
        vim.schedule(function() self.connect() end)
      end
    end)
  end

  function self:_on_data(data)
    self._buf = self._buf .. data
    -- Process all complete lines.
    while true do
      local nl = self._buf:find("\n", 1, true)
      if not nl then break end
      local line = self._buf:sub(1, nl - 1)
      self._buf  = self._buf:sub(nl + 1)
      if line ~= "" then self:_on_line(line) end
    end
  end

  function self:_on_line(line)
    local ok, msg = pcall(proto.decode_server, line)
    if not ok then
      self.log("malformed frame: " .. tostring(msg))
      return
    end
    vim.schedule(function() self:_dispatch(msg) end)
  end

  function self:_dispatch(msg)
    if msg.type == "hello_ack" then
      self.log(("HelloAck received (protocol v%d). Adapter ready."):format(
        msg.protocol_version or 0))
    elseif msg.type == "callback_request" then
      if self._handler then self._handler(msg.id, msg) end
    end
  end

  -- ── discovery file ────────────────────────────────────────────────────────

  function self:_read_discovery()
    local dir = disc_dir
    if vim.fn.isdirectory(dir) == 0 then return nil end
    local files = vim.fn.glob(dir .. "/*.json", false, true)
    if #files == 0 then return nil end

    -- Parse all valid, live entries.
    local entries = {}
    for _, path in ipairs(files) do
      local f = io.open(path, "r"); if not f then goto continue end
      local raw = f:read("*a"); f:close()
      local ok, obj = pcall(vim.json.decode, raw)
      if not ok or type(obj) ~= "table" or not obj.addr or not obj.pid then
        goto continue
      end
      -- Skip dead processes.
      if vim.loop.kill(obj.pid, 0) ~= 0 and
         not pcall(function() assert(vim.loop.kill(obj.pid, 0) == 0) end) then
        -- use os-level check instead
      end
      local host, port = obj.addr:match("^(.+):(%d+)$")
      if not host then goto continue end
      local mtime = (vim.loop.fs_stat(path) or {mtime={sec=0}}).mtime.sec
      -- Determine if parent is cade-server (preferred) or cade TUI (avoid).
      local ppid_out = vim.fn.system("ps -o ppid= -p " .. obj.pid .. " 2>/dev/null")
      local ppid = tonumber(ppid_out:match("%d+"))
      local parent_cmd = ppid and vim.fn.system(
        "ps -o comm= -p " .. ppid .. " 2>/dev/null"):gsub("%s","") or ""
      local is_server = parent_cmd:find("cade%-server") ~= nil
        or parent_cmd:find("cade_server") ~= nil
      table.insert(entries, {
        addr = host, port = tonumber(port), mtime = mtime,
        is_server = is_server, path = path,
      })
      ::continue::
    end

    if #entries == 0 then return nil end

    -- Prefer server-session entries; fall back to newest by mtime.
    table.sort(entries, function(a, b)
      if a.is_server ~= b.is_server then return a.is_server end
      return a.mtime > b.mtime
    end)

    local e = entries[1]
    return { addr = e.addr, port = e.port }
  end

  return self
end

return M
